Source/diablo.cpp

```diff
@@ -642,6 +727,12 @@ void GameEventHandler(const SDL_Event &event, uint16_t modState)
 			sgOptions.Keymapper.KeyReleased(event.button.button | KeymapperMouseButtonMask);
 		}
 		return;
+	case SDL_JOYBUTTONDOWN:
+		PressControllerButton(static_cast<ControllerButton>(event.jbutton.button));

```

Comment:
1. Why is this being handled here now?

2. This doesn't seem correct, `event.jbutton.button` and `ControllerButton` have different domains.
Isn't `event.jbutton.button` an arbitrary numeric code? We usually translate joystick events via `Joystick::ToControllerButton`.

Reply:
In a similar fashion to how we handle gamepad events as keystrokes, I modified `FetchMessage_Real()` to convert all gamepad events to `SDL_JOYBUTTON*` events specifically so they could be handled by the `GameEventHandler()`. In doing so, I capture the `ControllerButton` code as the `button` value for the converted event.

https://github.com/diasurgical/devilutionX/blob/beabeb0b8d43d94adaf53b890f931eca0c0c5145/Source/miniwin/misc_msg.cpp#L177-L179

As for why it was moved, it's because the padmapper actions should only run when `GameEventHandler` is registered. This is the same way that keymapper is used. I wouldn't say I'm particularly proud of the way this is handled, but the registered event handler can only handle SDL events so I figured this was a decent compromise.

Reply:
If a real `SDL_JOYBUTTONDOWN` comes in (e.g. the extra buttons on the Switch controller), wouldn't this cast be invalid?

Perhaps we should move the gamepad handling out of `FetchMessage_Real` to the actual event handling code (such as `GameEventHandler` etc)?

Reply:
No, that shouldn't happen. The `FetchMessage_Real()` function takes the `event` pointer from the caller and then captures the result of `PollEvent()` to the `e` variable. Data has to be explicitly written to the `event` variable that was passed in for it to reach the registered event handler. This means it will only send an `SDL_JOYBUTTONDOWN` event to the registered event handler if the `ControllerButton` is recognized.

It might indeed make some sense to move the rest of the event handling code to the `GameEventHandler`, but I felt that `FetchMessage_Real()` appears to be specifically designed for filtering and translating so to me it seemed more appropriate to leave it there. This way, the keystroke that the gamepad event gets translated into would be the one handled by the `GameEventHandler`.

Reply:
We ideally want to remove all the redundant translation layers and `misc_msg.cpp` in the future (I did some work towards that in https://github.com/diasurgical/devilutionX/issues/5025).

How about we make a separate function to do the gamepad-related stuff that is currently being done in `misc_msg.cpp` and call that from `GameEventHandler`?
Then we won't need the hacky `SDL_JOYBUTTON` at all!